engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,404 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/engine/lib"
+	"github.com/pingcap/tiflow/engine/pkg/p2p"
+	"github.com/pingcap/tiflow/pkg/workerpool"
+	"go.uber.org/zap"
+)
+
+var (
+	defaultMessageTimeOut  = time.Second * 2
+	defaultRequestTimeOut  = time.Second * 30
+	defaultResponseTimeOut = time.Second * 2
+	defaultHandlerTimeOut  = time.Second * 30
+)
+
+// generateTopic generate dm message topic.
+func generateTopic(senderID string, receiverID string) string {
+	return fmt.Sprintf("DM---%s---%s", senderID, receiverID)
+}
+
+// extractTopic extract dm message topic.
+// TODO: handle special case.
+func extractTopic(topic string) (string, string) {
+	parts := strings.Split(topic, "---")
+	return parts[1], parts[2]
+}
+
+type messageID uint64
+
+type messageType int
+
+const (
+	messageTp messageType = iota
+	requestTp
+	responseTp
+)
+
+// message use for asynchronous message and synchronous request/response.
+type message struct {
+	ID      messageID
+	Type    messageType
+	Command string
+	Payload interface{}
+}
+
+// messageIDAllocator is an id allocator for p2p message system
+type messageIDAllocator struct {
+	mu sync.Mutex
+	id messageID
+}
+
+// alloc allocs a new message id
+func (a *messageIDAllocator) alloc() messageID {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.id++
+	return a.id
+}
+
+// Sender defines an interface that supports send message
+type Sender interface {
```

Comment:
Can you say more on who should implement a `Sender` and where a `Sender` is supposed to be created?